### ISSUE TITLE
Normalize file paths/names.

### ISSUE THREAD

**Issue by klauspost** [rclone/rclone#195]
This isn't a huge issue, but we should probably normalize the files names when comparing them, otherwise encoding issues could lead to duplicate files/directories.

It seems like NFC (Canonical Decomposition, followed by Canonical Composition) is what we want.

NFKC is more intrusive, but can be useful in cases where the underlying filesystem does transforms like `[]rune{'ﬂ'}` → `[]rune{'f', 'l'}`. However, it also transforms `¼` → `1/4`. Here is [a list of more substitution examples](http://www.unicode.org/reports/tr15/).

It doesn't look like it is a huge task, but will need a few tests.

Here are the important changes I would by looking for usage of the `Fs.Remote()` interface value, which to me indicates operations across filesystems.

If you think this is the right way to go, I will see if I can create some sensible tests.

``` diff
diff --git a/fs/operations.go b/fs/operations.go
index 7b8fc86..e537415 100644
--- a/fs/operations.go
+++ b/fs/operations.go
@@ -10,6 +10,8 @@ import (
    "sync"
    "sync/atomic"
    "time"
+
+   "code.google.com/p/go.text/unicode/norm"
 )

 // CalculateModifyWindow works out modify window for Fses passed in -
@@ -182,7 +184,7 @@ tryAgain:
    actionTaken := "Copied (server side copy)"
    if fCopy, ok := f.(Copier); ok && src.Fs().Name() == f.Name() {
        var newDst Object
-       newDst, err = fCopy.Copy(src, src.Remote())
+       newDst, err = fCopy.Copy(src, norm.NFC.String(src.Remote()))
        if err == nil {
            dst = newDst
        }
@@ -211,7 +213,7 @@ tryAgain:
            err = dst.Update(in, src.ModTime(), src.Size())
        } else {
            actionTaken = "Copied (new)"
-           dst, err = f.Put(in, src.Remote(), src.ModTime(), src.Size())
+           dst, err = f.Put(in, norm.NFC.String(src.Remote()), src.ModTime(), src.Size())
        }
        inErr = in.Close()
    }
@@ -338,7 +340,7 @@ func PairMover(in ObjectPairChan, fdst Fs, wg *sync.WaitGroup) {
                    ErrorLog(dst, "Couldn't delete: %v", err)
                }
            }
-           _, err := fdstMover.Move(src, src.Remote())
+           _, err := fdstMover.Move(src, norm.NFC.String(src.Remote()))
            if err != nil {
                Stats.Error()
                ErrorLog(dst, "Couldn't move: %v", err)
@@ -384,7 +386,7 @@ func DeleteFiles(toBeDeleted ObjectsChan) {
 func readFilesMap(fs Fs) map[string]Object {
    files := make(map[string]Object)
    for o := range fs.List() {
-       remote := o.Remote()
+       remote := norm.NFC.String(o.Remote())
        if _, ok := files[remote]; !ok {
            files[remote] = o
        } else {
@@ -444,7 +446,7 @@ func syncCopyMove(fdst, fsrc Fs, Delete bool, DoMove bool) error {

    go func() {
        for src := range fsrc.List() {
-           remote := src.Remote()
+           remote := norm.NFC.String(src.Remote())
            dst, dstFound := delFiles[remote]
            if !Config.Filter.Include(remote, src.Size()) {
                Debug(src, "Excluding from sync")
```

``` diff
diff --git a/fs/limited.go b/fs/limited.go
index 3c15ba2..b94c677 100644
--- a/fs/limited.go
+++ b/fs/limited.go
@@ -4,6 +4,8 @@ import (
    "fmt"
    "io"
    "time"
+
+   "golang.org/x/text/unicode/norm"
 )

 // Limited defines a Fs which can only return the Objects passed in
@@ -58,8 +60,9 @@ func (f *Limited) ListDir() DirChan {

 // NewFsObject finds the Object at remote.  Returns nil if can't be found
 func (f *Limited) NewFsObject(remote string) Object {
+   remote = norm.NFC.String(remote)
    for _, obj := range f.objects {
-       if obj.Remote() == remote {
+       if norm.NFC.String(obj.Remote()) == remote {
            return obj
        }
    }
@@ -72,6 +75,7 @@ func (f *Limited) NewFsObject(remote string) Object {
 // will return the object and the error, otherwise will return
 // nil and the error
 func (f *Limited) Put(in io.Reader, remote string, modTime time.Time, size int64) (Object, error) {
+   remote = norm.NFC.String(remote)
    obj := f.NewFsObject(remote)
    if obj == nil {
        return nil, fmt.Errorf("Can't create %q in limited fs", remote)
@@ -106,6 +110,7 @@ func (f *Limited) Precision() time.Duration {
 //
 // If it isn't possible then return fs.ErrorCantCopy
 func (f *Limited) Copy(src Object, remote string) (Object, error) {
+   remote = norm.NFC.String(remote)
    fCopy, ok := f.fs.(Copier)
    if !ok {
        return nil, ErrorCantCopy

```

**Comment by ncw** [comment]
I'm assuming that this is a fix for #194?

If so then it would be a lot less ugly if we could confine it to the local file system, and assume that the remotes will be sensible and just return whatever utf-8 we give them, rather than mangling it like OS X and HFS+ do?

**Comment by klauspost** [comment]
No. I actually didn't know that the root cause for #194 was normalization.

Assume we sync a file from FS A -> B. When we read back the directory of B, we get it back in a different form. The next time we sync A -> B, the file will not be found since the representation differ, and therefore re-sync'ed every time.

This compares the normalized form of the files.

That said, we should probably give each FS a change to do individual normalization (like the Windows local file system), but this should still be here.

**Comment by zeshanb** [comment]
Unfortunately, the best option to copy an entire drive or part of a volume
in OS X is to run Disk Utility in repair boot:

http://cdn.osxdaily.com/wp-content/uploads/2012/02/repair-disk-disk-utility.jpg

This is specially the case for external storage depending on your system
config. The OS overhead and other background processes can weigh down on
transfers over USB.

**Comment by ncw** [comment]
@klauspost Been thinking about this...

I think you are right - the only way to do this is in the copy operations, at the interface between one file system and the other.

The alternative would be to say that all remotes must output normalised file names.  We could do this, but for file systems that have unnormalised file names (like HFS+ on OS X) we'd need to keep a copy of the internal unnormalised name and the `Remote` name and be able to translate between the two which gets very complicated very quickly. This doesn't cover the case where someone might have uploaded an unnormalised file name to S3 say and there is no easy way to get the unnormalised name from the normalised name.

So yes, it would be great if you have a go with this.

I'd probably make some utility function just to try to reduce the visual complexity of the change.

```go
func NormString(s string) string {
  return norm.NFC.String(s)
}

func NormRemote(o Object) string {
  return NormString(o.Remote())
}
```  

I'm pretty sure that this will fix #194

**Comment by ivandeex** [comment]
Latest rclone encodes filenames. This issue needs a retest.

**Comment by ncw** [comment]
What we do now is
- don't normalize file names in the fs.Object-s - this caused too much trouble otherwise
- do compare filenames with normalization in the sync routines

So I think this issue has been fixed by 4ac9a65049581d20466a6426856297d597df28f5 and #1477 where this issue is discussed more.
